tools/unicode/c/genprops/layoutpropsbuilder.cpp

```diff
@@ -156,126 +239,62 @@ void LayoutPropsBuilder::checkMaxIntValue(UProperty prop, int32_t maxMax,
     }
 }
 
+// In ICU 63, we had functions writeCSourceFile() and writeJavaSourceFile().
+// For Java, each serialized trie was written as a String constant with
+// one byte per char and an optimization for byte 0,
+// to optimize for Java .class file size.
+// (See ICU 63 if we need to resurrect some of that code.)
+// Since ICU 64, we write a binary ulayout.icu file for use in both C++ & Java.
+
 void
-LayoutPropsBuilder::writeCSourceFile(const char *path, UErrorCode &errorCode) {
+LayoutPropsBuilder::writeBinaryData(const char *path, UBool withCopyright, UErrorCode &errorCode) {
     if (U_FAILURE(errorCode)) { return; }
 
-    FILE *f = usrc_create(path, "ulayout_props_data.h", 2018,
-                          "icu/tools/unicode/c/genprops/layoutpropsbuilder.cpp");
-    if (f == nullptr) {
-        errorCode = U_FILE_ACCESS_ERROR;
+    UNewDataMemory *pData = udata_create(
+        path, ULAYOUT_DATA_TYPE, ULAYOUT_DATA_NAME, &dataInfo,
+        withCopyright ? U_COPYRIGHT_STRING : NULL, &errorCode);
+    if (U_FAILURE(errorCode)) {
+        fprintf(stderr, "genprops: udata_create(%s, ulayout.icu) failed - %s\n",
+                path, u_errorName(errorCode));
         return;
     }
-    fputs("#ifdef INCLUDED_FROM_UPROPS_CPP\n\n", f);
-
-    writeMaxIntValue(f, "InPC", UCHAR_INDIC_POSITIONAL_CATEGORY);
-    usrc_writeUCPTrie(f, "inpc", inpcTrie);
-
-    writeMaxIntValue(f, "InSC", UCHAR_INDIC_SYLLABIC_CATEGORY);
-    usrc_writeUCPTrie(f, "insc", inscTrie);
 
-    writeMaxIntValue(f, "Vo", UCHAR_VERTICAL_ORIENTATION);
-    usrc_writeUCPTrie(f, "vo", voTrie);
+    int32_t indexes[ULAYOUT_IX_COUNT] = { ULAYOUT_IX_COUNT };
+    int32_t top = ULAYOUT_IX_COUNT * 4;
 
-    fputs("#endif  // INCLUDED_FROM_UPROPS_CPP\n", f);
-    fclose(f);
-}
-
-namespace {
+    indexes[ULAYOUT_IX_INPC_TRIE_TOP] = top += inpcLength;
```

Comment:
I had to loop up the operator precedence for `+=` and `=`.  Maybe use parentheses to be more clear?

Reply:
done